### ISSUE TITLE
Make paser more powerful

### ISSUE THREAD

**Issue by wjhuang2016** [pingcap/tidb#10168]
## Feature Request

**Is your feature request related to a problem? Please describe:**

When pasering a statement, we only turn the info into a special ast presently. For example, a `select` statement will return a selectStmt.
However, we may need to get some infomation in paser stage which will present in many statement.  If we add these infomation in every ast which need it, the code can not be concise.
In postgresql, the paser will return a common selectStmt. In MySQL, a struct `LEX` will save the common info.

Here a problem that need a more powerful parser:
In #10004, a concept `constItem` is introduced. There are some errors in handling nondeterministic functions. For 'getVar` func, if the val is set in the same statement, the func is nondeterministic. Otherwise, it's deterministic.
```
mysql> select @a:=FROM_UNIXTIME(1), unix_timestamp(@a);
+----------------------+--------------------+
| @a:=FROM_UNIXTIME(1) | unix_timestamp(@a) |
+----------------------+--------------------+
| 1970-01-01 08:00:01  |           1.000000 |
+----------------------+--------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select @a:=FROM_UNIXTIME(1);
+----------------------+
| @a:=FROM_UNIXTIME(1) |
+----------------------+
| 1970-01-01 08:00:01  |
+----------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select unix_timestamp(@a);
+--------------------+
| unix_timestamp(@a) |
+--------------------+
|                  1 |
+--------------------+
1 row in set (0.00 sec)
```
So we need to know all the set statements in a statement and It's the job of parser.

**Describe the feature you'd like:**

Let paser return more infomation
**Describe alternatives you've considered:**

Let paser return a extra struct?
**Teachability, Documentation, Adoption, Migration Strategy:**

**Comment by zz-jason** [comment]
> So we need to know all the set statements in a statement and It's the job of parser.

The parser has already provided the `set` statement as `SetVar` function. Could you post the the problems you encountered?

**Comment by wjhuang2016** [comment]
@zz-jason 
Yes, the parser has provided `SetVar` function, and I want to get all the `Set statement` in `StatementContext`. To get it I have to check if there is a `SetVar` function recursively, and I think it would be great that the parser return a `StatementContext` like struct.
The problem I encoutered is:

In this command:
`mysql> select @a:=FROM_UNIXTIME(1), unix_timestamp(@a);`
`unix_timestamp` should not be a `constItem`, Because the `a` is set in the same statement. 

In following commands:
```
select @a:=FROM_UNIXTIME(1);
select unix_timestamp(@a);
```
`unix_timestamp` should be a `constItem`.

To solve the above problem, we need to get all the `Set statement` in `StatementContext`
